torch/_inductor/fx_passes/pad_mm.py

```diff
@@ -205,35 +256,61 @@ def tensor_key(t):
 def should_pad_bench(
     mat1: Tensor, mat2: Tensor, op, input: Optional[Tensor] = None
 ) -> bool:
-    if not has_triton():
-        return False
+    if torch._inductor.config.shape_pad_always:
+        return True
 
     do_bench = functools.partial(
         utils.do_bench,
         warmup=5,
     )
-
+    m_padded_length = 0
+    n_padded_length = 0
+    batchsize = 1
+    explicit_transpose = False
     with no_dispatch():
         if op is torch.ops.aten.mm or op is torch.ops.aten.addmm:
             m = mat1.shape[0]
             k = mat1.shape[1]
             n = mat2.shape[1]
-
-            m_padded_length = get_padded_length(m, get_alignment_size(mat1))
             k_padded_length = get_padded_length(k, get_alignment_size(mat1))
-            n_padded_length = get_padded_length(n, get_alignment_size(mat2))
+            if not torch._inductor.config.shape_pad_only_k_dim:
+                n_padded_length = get_padded_length(n, get_alignment_size(mat2))
+                m_padded_length = get_padded_length(m, get_alignment_size(mat1))
         elif op is torch.ops.aten.bmm:
+            batchsize = mat1.shape[0]
             m = mat1.shape[1]
-            k = mat2.shape[2]
+            k = mat1.shape[2]
             n = mat2.shape[2]
-
-            m_padded_length = get_padded_length(m, get_alignment_size(mat1))
             k_padded_length = get_padded_length(k, get_alignment_size(mat1))
-            n_padded_length = get_padded_length(n, get_alignment_size(mat2))
+            if not torch._inductor.config.shape_pad_only_k_dim:
+                m_padded_length = get_padded_length(m, get_alignment_size(mat1))
+                n_padded_length = get_padded_length(n, get_alignment_size(mat2))
         else:
             return False
 
-        if m_padded_length == k_padded_length == n_padded_length == 0:
+        if torch._inductor.config.shape_pad_use_transpose:
+            if m_padded_length == 0 and n_padded_length != 0:
+                n_padded_length = 0
+                m_padded_length = 0
+                explicit_transpose = True
+            elif n_padded_length == 0 and m_padded_length != 0:
+                m_padded_length = 0
+        if (
+            m_padded_length == k_padded_length == n_padded_length == 0
+        ) and not explicit_transpose:
+            return False
+
+        fake_layout = FixedLayout(
+            device=mat1.device,
+            dtype=mat1.dtype,
+            size=[batchsize, m, n],
+            stride=[n * m, n, 1],
+        )
+        if use_cutlass_template(fake_layout):
```

Comment:
Do you still want to pad bandwidth bound mms in cutlass ? I found for the small matmuls it wasnt profitable to pad

Reply:
We're not going to use cutlass for small Matmuls, in practice they will have a size threshold (in terms of M*N*K) below which they won't be used. This is how I did it here ( Meta-internal link) https://fb.workplace.com/notes/347669491400529/ and arrived at whole-model speedups of up to 14%.